### PR TITLE
Use SIMD in fast path for advance_for_char_range

### PR DESCRIPTION
**Pull request by bjwbell**
In advance_for_char_range add a fast SIMD code path for the the common
case where there are no detailed glyphs.

r? @mbrubeck

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7527)

### CODE REVIEW

File: components/gfx/text/glyph.rs [hunk 1]
```diff
@@ -541,8 +547,41 @@ impl<'a> GlyphStore {
...
+
+    #[inline]
+    fn advance_for_char_range_simple_glyphs(&self, rang: &Range<CharIndex>) -> Au {
+        let mask = u32x4::splat(GLYPH_ADVANCE_MASK);
+        let mut simd_advance = u32x4::splat(0);
+        let begin = rang.begin().to_usize();
+        let len = rang.length().to_usize();
+        let num_simd_iterations = len / 4;
+        let leftover_entries = rang.end().to_usize() - (len - num_simd_iterations * 4);
+        let buf: &[u32] = unsafe { mem::transmute(self.entry_buffer.as_slice()) };
```
Comment (pcwalton) [inline components/gfx/text/glyph.rs]: Can you factor this unsafe call out into a separate method and add a comment to the entry buffer field noting that if it is changed you may need to change the method too?

The safer way would be to use something like the nue crate here, but let's not pull in that dependency yet.

Comment (bjwbell) [inline components/gfx/text/glyph.rs]: Sorry for the delay. I factored it out today.

**Comment by bjwbell** [comment]
Looks like the SIMD crate fails to build on android due to an unused import (arm::neon::common). I'll look at it later today (have to run out for couple hours).

**Comment by huonw** [comment]
The problem here is NEON isn't guaranteed to exist on all ARM devices, meaning `rustc` (and hence `simd`) cannot assume it exists for the default `arm-...` targets. As you've found out, the crate handles these cases particularly badly. (It also affects 32-bit x86 (the `i686-...`) targets at the moment.)

One way to fix is to use a wrapper script around rustc (e.g. like `etc/rustc-with-gold`) that passes `-C target-feature=+neon` on the arm targets, if you're OK with only running on devices with NEON support.

**Comment by bjwbell** [comment]
For now, I'll disable SIMD in servo on ARM. And put in a note in the code that for ARM it'll always use the slow path.  

I'd rather do the wrapper script as a separate PR, since I've never modified the servo build process.

**Comment by bjwbell** [comment]
Having https://github.com/rust-lang/cargo/issues/1137 (Expose target-cpu/target-feature) would be really nice for this.

**Comment by bjwbell** [comment]
@pcwalton can you have bors rebuild & test? I can't do it locally (only have an x86_64).

**Comment by bjwbell** [comment]
Third times the charm.

I changed the simd dependency in gfx/Cargo.toml to be target triple specific and added target triples
for:
- x86_64-unknown-linux-gnu
- x86_64-apple-darwin
- aarch64-unknown-linux-gnu
- x86_64-pc-windows-gnu
- x86_64-pc-windows-msvc

No idea, if the android builder's target triple is in the above list. 

@pcwalton you can retry bors.

**Comment by metajack** [comment]
The android triple is arm-linux-androideabi

**Comment by bjwbell** [comment]
@metajack thanks, I realized I only need it for aarch64 on android since the simd crate isn't used for any of the other arm android platforms.

**Comment by bjwbell** [comment]
@metajack Is the android aarch64 triple aarch64-linux-androideabi? If it is I can add it.

**Comment by metajack** [comment]
@bjwbell I don't actually know. @larsbergstrom ?

**Comment by metajack** [comment]
@bors-servo r+

---

Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 1 of 1 files at r5.
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7527)

**Comment by larsbergstrom** [comment]
@metajack @bjwbell Yes, that appears correct. For SIMD/NEON, I think we also need to figure out if we should be targeting `armv7-linux-androideabi`, since NEON is available in armv7, too, IIRC.

**Pull request merged**